### PR TITLE
use twemoji

### PR DESCRIPTION
**Pull request by sijad**
fixes flarum/core#1460

### CODE REVIEW

File: js/src/forum/addComposerAutocomplete.js [hunk 1]
```diff
@@ -11,6 +12,16 @@ export default function addComposerAutocomplete() {
 
   const emojiKeys = Object.keys(emojiMap);
 
+  function getEmojiIconCode(emoji) {
```
Comment (sijad) [inline js/src/forum/addComposerAutocomplete.js]: using only `toCodePoint()` breaks some emojis and twemoji does not export [this function](https://github.com/twitter/twemoji/blob/0dd7c46f41837f1aff7ae930e1993e20d66ddd62/2/twemoji.npm.js#L322) so I had to recreate it.

should we use something like this for emojimap instead:

```js
[
	{
		"emoji": "😂 ",
		"code": "1f602",
		"aliases": ["alias_1", "alias_2", "alias_3"],
	}
	// ...
]
```

Comment (tobscure) [reply js/src/forum/addComposerAutocomplete.js]: If recreating the twemoji function is reliable, then no need to pollute the emoji map with extra information?

Comment (sijad) [reply js/src/forum/addComposerAutocomplete.js]: I think it's working fine

**Comment by sijad** [comment]
I need to clean it up a little bit and I guess it's good to go. Should i make picker async in this pr?

**Comment by tobscure** [comment]
Nah do that in another PR

**Comment by sijad** [comment]
I'm using this script to [build emojiMap](https://gist.github.com/sijad/d2756a32e61bdc1b94d694d374817a5d#file-build_emoji_from_gemoji-php)
gemoji db contains `1432` emoji,
but twemoji does not work well with all of them (e.g. 👯‍♂️, 👨‍⚕️), so they had to be removed.

after remove those emojis, `emojiMap` contains `1265` emoji.

**Comment by sijad** [comment]
I guess removing those is not a good idea, I'm trying to find a better way to exclude these emojis.

**Comment by sijad** [comment]
now it seems alright, I'm using [this script](https://gist.github.com/sijad/d2756a32e61bdc1b94d694d374817a5d#file-build_emoji_from_gemoji-js) now which make sure all emoji will exist in twemoji.

does twemoji default settings is good? (svg, size, etc)

**Comment by tobscure** [comment]
@sijad looking good! Which emojis are not found, and do they actually exist in twemoji? For example, with 👯‍♂️ that you mentioned above, the unicode according to macOS is `U+1F46F U+200D U+2642 U+FE0F`, which [appears to exist in twemoji](https://github.com/twitter/twemoji/blob/gh-pages/2/72x72/1f46f-200d-2642-fe0f.png).

**Comment by sijad** [comment]
weird, it seems `toCodePoint()` ignores `U+FE0F`

currently my script can find these:

```
not found: 1f471-200d-2640 👱‍♀
not found: 1f473-200d-2640 👳‍♀
not found: 1f46e-200d-2640 👮‍♀
not found: 1f477-200d-2640 👷‍♀
not found: 1f482-200d-2640 💂‍♀
not found: 1f469-200d-2695 👩‍⚕
not found: 1f468-200d-2695 👨‍⚕
not found: 1f469-200d-2708 👩‍✈
not found: 1f468-200d-2708 👨‍✈
not found: 1f469-200d-2696 👩‍⚖
not found: 1f468-200d-2696 👨‍⚖
not found: 1f647-200d-2640 🙇‍♀
not found: 1f481-200d-2642 💁‍♂
not found: 1f645-200d-2642 🙅‍♂
not found: 1f646-200d-2642 🙆‍♂
not found: 1f64b-200d-2642 🙋‍♂
not found: 1f926-200d-2640 🤦‍♀
not found: 1f926-200d-2642 🤦‍♂
not found: 1f937-200d-2640 🤷‍♀
not found: 1f937-200d-2642 🤷‍♂
not found: 1f64e-200d-2642 🙎‍♂
not found: 1f64d-200d-2642 🙍‍♂
not found: 1f487-200d-2642 💇‍♂
not found: 1f486-200d-2642 💆‍♂
not found: 1f46f-200d-2642 👯‍♂
not found: 1f6b6-200d-2640 🚶‍♀
not found: 1f3c3-200d-2640 🏃‍♀
not found: 1f93c-200d-2640 🤼‍♀
not found: 1f93c-200d-2642 🤼‍♂
not found: 1f938-200d-2640 🤸‍♀
not found: 1f938-200d-2642 🤸‍♂
not found: 1f93e-200d-2640 🤾‍♀
not found: 1f93e-200d-2642 🤾‍♂
not found: 1f3c4-200d-2640 🏄‍♀
not found: 1f3ca-200d-2640 🏊‍♀
not found: 1f93d-200d-2640 🤽‍♀
not found: 1f93d-200d-2642 🤽‍♂
not found: 1f6a3-200d-2640 🚣‍♀
not found: 1f6b4-200d-2640 🚴‍♀
not found: 1f6b5-200d-2640 🚵‍♀
not found: 1f939-200d-2640 🤹‍♀
not found: 1f939-200d-2642 🤹‍♂
```

**Comment by sijad** [comment]
it seems gemoji db does not does not contains `U+FE0F`, when I copy `dancing_men` from gemoji json twemoji render it like this:

image

when I copy it from somewhere it does render correctly, am I missing something?

**Comment by sijad** [comment]
ok, new script fixed that. not it tries to add `U+FE0F` and check again if an emoji could not found, now every gemoji have found.

**Comment by franzliedke** [comment]
@sijad I hope you don't mind if we delay reviewing / discussing this until after beta-8. Full focus on that release now. :steam_locomotive:

**Comment by sijad** [comment]
https://github.com/flarum/core/pull/1457 breaks this extension AFAIK

**Comment by tobscure** [comment]
@franzliedke we may need to merge this for beta 8, because having upgraded TextFormatter we've introduced licensing issues. See https://github.com/flarum/core/issues/1460

**Comment by tobscure** [comment]
@sijad I think our best bet going forward is to use emojibase. The emojiMap generated by your script wasn't quite working, I fixed it as you can see here: https://gist.github.com/tobscure/d409f787d719d5cfcfedbca1fa31204f/revisions#diff-168726dbe96b3ce427e7fedce31bb0bc

Also, we only want to include `shortcodes` in the map.

Have confirmed the output is working nicely with this PR.

Can you please make this script a part of flarum-ext-emoji. I reckon add it to `js/scripts/generate-emoji-map.js`, add emojibase to the package.json, add an npm script called `generate-emoji-map` that runs it (and call that script at the start of the `build` script), and make the script output to `src/forum/generated/emojiMap.js`.

**Comment by tobscure** [comment]
As far as I could tell they weren't included

**Comment by sijad** [comment]
you're right

by using your changes it produce this error:

`Can not find 👁️‍🗨️ 1F441-FE0F-200D-1F5E8-FE0F`

am I missing something here?

**Comment by tobscure** [comment]
Yeah that's the only one it can't find. I'm not sure why? But otherwise it works perfectly :)

File: js/src/forum/generated/emojiMap.json [hunk 2]
```diff
@@ -0,0 +1 @@
...],"🇷🇴":["flag_ro"],"🇷🇸":["flag_rs"],"🇷🇺":["flag_ru"],"🇷🇼":["flag_rw"],"🇸🇦":["flag_sa"],"🇸🇧":["flag_sb"],"🇸🇨":["flag_sc"],"🇸🇩":["flag_sd"],"🇸🇪":["flag_se"],"🇸🇬":["flag_sg"],"🇸🇭":["flag_sh"],"🇸🇮":["flag_si"],"🇸🇯":["flag_sj"],"🇸🇰":["flag_sk"],"🇸🇱":["flag_sl"],"🇸🇲":["flag_sm"],"🇸🇳":["flag_sn"],"🇸🇴":["flag_so"],"🇸🇷":["flag_sr"],"🇸🇸":["flag_ss"],"🇸🇹":["flag_st"],"🇸🇻":["flag_sv"],"🇸🇽":["flag_sx"],"🇸🇾":["flag_sy"],"🇸🇿":["flag_sz"],"🇹🇦":["flag_ta"],"🇹🇨":["flag_tc"],"🇹🇩":["flag_td"],"🇹🇫":["flag_tf"],"🇹🇬":["flag_tg"],"🇹🇭":["flag_th"],"🇹🇯":["flag_tj"],"🇹🇰":["flag_tk"],"🇹🇱":["flag_tl"],"🇹🇲":["flag_tm"],"🇹🇳":["flag_tn"],"🇹🇴":["flag_to"],"🇹🇷":["flag_tr"],"🇹🇹":["flag_tt"],"🇹🇻":["flag_tv"],"🇹🇼":["flag_tw"],"🇹🇿":["flag_tz"],"🇺🇦":["flag_ua"],"🇺🇬":["flag_ug"],"🇺🇲":["flag_um"],"🇺🇳":["flag_un"],"🇺🇸":["flag_us","usa"],"🇺🇾":["flag_uy"],"🇺🇿":["flag_uz"],"🇻🇦":["flag_va"],"🇻🇨":["flag_vc"],"🇻🇪":["flag_ve"],"🇻🇬":["flag_vg"],"🇻🇮":["flag_vi"],"🇻🇳":["flag_vn"],"🇻🇺":["flag_vu"],"🇼🇫":["flag_wf"],"🇼🇸":["flag_ws"],"🇽🇰":["flag_xk"],"🇾🇪":["flag_ye"],"🇾🇹":["flag_yt"],"🇿🇦":["flag_za"],"🇿🇲":["flag_zm"],"🇿🇼":["flag_zw"],"🏴󠁧󠁢󠁥󠁮󠁧󠁿":["flag_gbeng","england"],"🏴󠁧󠁢󠁳󠁣󠁴󠁿":["flag_gbsct","scotland"],"🏴󠁧󠁢󠁷󠁬󠁳󠁿":["flag_gbwls","wales"]}
```
Comment (tobscure) [inline js/src/forum/generated/emojiMap.json]: This file probably doesn't need to be committed? Since it is generated when building

**Comment by sijad** [comment]
I've used `.json` instead, is that ok?

**Comment by tobscure** [comment]
Yep .json is great

File: js/scripts/generate-emoji-map.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
+const fs = require('fs');
+const path = require('path');
+const twemoji = require('twemoji/2/twemoji.npm');
+
+const data = require('emojibase-data/en/data.json');
+const twemojiFileNames = fs.readdirSync('./node_modules/twemoji/2/svg')
+  .map(name => path.basename(name, '.svg'));
+
+const alternative = {
```
Comment (sijad) [inline js/scripts/generate-emoji-map.js]: please see https://github.com/twitter/twemoji/issues/253
emojibase follows http://unicode.org/Public/emoji/ so it's not emojibase problem, but it seems `👁‍🗨` represented differently in various platforms, for instance macOS's emoji picker insert it without `FE0F`, so I'm also not sure if it's twemoji problem either. 

we can use this hack or we should ignore `👁‍🗨`.

Comment (tobscure) [reply js/scripts/generate-emoji-map.js]: That hack is fine :)

File: js/src/forum/generated/emojiMap.json (hunk 2 above)
Comment (sijad) [reply js/src/forum/generated/emojiMap.json]: should I add `mkdir src/forum/generated` to build command or should I add `.gitkeep`?

Comment (sijad) [reply js/src/forum/generated/emojiMap.json]: or maybe script should make that folder too?

File: js/scripts/generate-emoji-map.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  const emojiCode = getEmojiIconCode(emoji);
+
+  if (!checkExistanceInTwemoji(emojiCode)) {
+    console.error('Can not find', emoji, emojiCode);
+    continue;
+  }
+
+  emojis[emoji] = [
+    ...shortcodes,
+  ];
```
Comment (tobscure) [inline js/scripts/generate-emoji-map.js]: This can be simplified to `emojis[emoji] = shortcodes`

File: js/scripts/generate-emoji-map.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+};
+
+const emojis = {};
+
+for (let e of data) {
+  const {
+    name,
+    shortcodes,
+    tags,
+  } = e;
```
Comment (tobscure) [inline js/scripts/generate-emoji-map.js]: This can be simplified to const { shortcodes } = e

Comment (sijad) [reply js/scripts/generate-emoji-map.js]: and I wish for eslint

File: js/package.json [hunk 6]
```diff
@@ -2,13 +2,16 @@
...
+    "twemoji": "^11.0.0",
     "webpack": "^4.0.0",
     "webpack-cli": "^3.0.7"
   },
   "scripts": {
-    "build": "webpack --mode production",
-    "watch": "webpack --mode development --watch"
+    "build": "npm run generate-emoji-map; webpack --mode production",
+    "watch": "npm run generate-emoji-map; webpack --mode development --watch",
+    "generate-emoji-map": "node ./scripts/generate-emoji-map.js"
```
Comment (franzliedke) [inline js/package.json]: Could we use `prebuild` and `prewatch` here, so that we are more in line with other extensions?

**Comment by datitisev** [comment]
I mentioned this to Toby, but for us to have an open discussion, I'll post it here.

I made an NPM package called `simple-emoji-map` for reflar/reactions which could be used here to replace the emoji map. The build code comes right out of reflar/reactions, which changed some things from this PR's attempt.

I saw there was a comment on retrieving the code point of the emoji. With `simple-emoji-map` you can select whether to use `emoji` or `codepoint` for the JSON keys. The emoji is easy to get from the codepoint, as opposed from getting the emoji from the codepoint:
```js
const toUnicodeEmoji = codePoint => String.fromCodePoint(...codePoint.split('-').map(e => `0x${e}`));
```

Using `codepoint` for the keys only adds 4KB, which I don't think would really make a different. Even if you continue to use emoji, I think using `simple-emoji-map` would simplify things.

<details> 
  <summary>Simple-Emoji-Map</summary>

`simple-emoji-map` builds the json file on `install` (and it can be rebuilt manually with `npm rebuild simple-emoji-map`). The configuration is loaded on every build, so everyone should end up with the same map when compiling.

You can learn more about it at [datitisev/simple-emoji-map](https://github.com/datitisev/simple-emoji-map)

  image
</details>

**Comment by sijad** [comment]
`simple-emoji-map` seems to be `generate-emoji-map.js`, right?
I'm not fan of adding dependencies, but I'm not against it either, it's Toby and Franz calls, please let me know what you think?

**Comment by datitisev** [comment]
@sijad Yep! I got it from this PR and changes I made for reflar/reactions, and made it use the config so you can customize it.
And as it builds the map file when it's installed/built, it doesn't run every time `build` or `watch` runs (even though that doesn't take long anyway).
And generating the map is as easy as running `npm rebuild simple-emoji-map`, using the latest `emojibase-data@^3.1.0` and `twemoji@^11.0.1`

**Comment by franzliedke** [comment]
@datitisev @sijad Thanks for your work on this, it's highly appreciated!

Please understand we won't spend more time on this before beta.8, in order to release as soon as possible.

**Comment by tobscure** [comment]
@franzliedke se https://github.com/flarum/flarum-ext-emoji/pull/16#issuecomment-406444033, I would be happy to merge as-is

**Comment by franzliedke** [comment]
Hmm, okay, I'll have a look soon-ish.

**Comment by franzliedke** [comment]
@sijad Thank you so much!

And sorry for asking twice whether we can delay this, even though all the information was already there. :see_no_evil:

**Comment by franzliedke** [comment]
@sijad @tobscure So does this fix flarum/core#1331 as well?

**Comment by datitisev** [comment]
@franzliedke Think so, and possibly https://github.com/flarum/core/issues/1297 as well.

**Pull request merged**